core/kernel/interrupt.c

```diff
@@ -37,51 +40,196 @@
  */
 
 static struct itr_chip *itr_chip;
-static SLIST_HEAD(, itr_handler) handlers = SLIST_HEAD_INITIALIZER(handlers);
+static struct itr_desc *itr_descs[256] = {0};
+static unsigned itr_lock;
 
 void itr_init(struct itr_chip *chip)
 {
 	itr_chip = chip;
 }
 
-static struct itr_handler *find_handler(size_t it)
-{
-	struct itr_handler *h;
-
-	SLIST_FOREACH(h, &handlers, link)
-		if (h->it == it)
-			return h;
-	return NULL;
-}
-
 void itr_handle(size_t it)
 {
-	struct itr_handler *h = find_handler(it);
+	struct itr_desc *desc = itr_descs[it];
 
-	if (!h) {
+	if (!desc) {
 		EMSG("Disabling unhandled interrupt %zu", it);
 		itr_chip->ops->disable(itr_chip, it);
 		return;
 	}
 
-	if (h->handler(h) != ITRR_HANDLED) {
+	/* if interrupt type is shared, should traverse all the handlers */
+	if ((desc->flags & ITRF_SHARE_MASK) == ITRF_SHARE) {
+		for (; desc; desc = desc->d_next) {
+			if (desc->handler(it, desc->data) == ITRR_HANDLED)
+				return;
+		}
+	} else if (desc->handler(it, desc->data) != ITRR_HANDLED) {
 		EMSG("Disabling interrupt %zu not handled by handler", it);
 		itr_chip->ops->disable(itr_chip, it);
 	}
 }
 
-void itr_add(struct itr_handler *h)
+static itr_result add_handler(size_t it, itr_handler handler,
+		uint32_t flags, void *data)
+{
+	itr_result res;
+	struct itr_desc *desc = itr_descs[it];
+	struct itr_desc *in_desc;
+	struct itr_desc *tmp_desc;
+	uint32_t old_itr_status;
+
+	in_desc = malloc(sizeof(struct itr_desc));
```

Comment:
It is best to zero-init the full structure. it prevents from forgetting to initialize each field individually.  use either `malloc();memset();` or `calloc()`.